stdlib/LibGit2/src/callbacks.jl

```diff
@@ -359,19 +359,131 @@ function fetchhead_foreach_callback(ref_name::Cstring, remote_url::Cstring,
     return Cint(0)
 end
 
+function verify_host_error(message::AbstractString)
+    printstyled(stderr, "$message\n", color = :cyan, bold = true)
+end
+
 function certificate_callback(
     cert_p :: Ptr{Cvoid},
     valid  :: Cint,
     host_p :: Ptr{Cchar},
-    user_p :: Ptr{Cvoid},
+    data_p :: Ptr{Cvoid},
 )::Cint
     valid != 0 && return Consts.CERT_ACCEPT
     host = unsafe_string(host_p)
     cert_type = unsafe_load(convert(Ptr{Cint}, cert_p))
     transport = cert_type == Consts.CERT_TYPE_TLS ? "TLS" :
                 cert_type == Consts.CERT_TYPE_SSH ? "SSH" : nothing
-    verify = NetworkOptions.verify_host(host, transport)
-    verify ? Consts.PASSTHROUGH : Consts.CERT_ACCEPT
+    if !NetworkOptions.verify_host(host, transport)
+        # user has opted out of host verification
+        return Consts.CERT_ACCEPT
+    end
+    if transport == "TLS"
+        # TLS verification is done before the callback
+        # so if we get here, host verification failed
+        verify_host_error("TLS host verification: the identity of the server `$host` could not be verified. Someone could be trying to man-in-the-middle your connection. It is also possible that the correct server is using an invalid certificate or that your system's certificate authority root store is misconfigured.")
+        return Consts.CERT_REJECT
+    end
+    if transport == "SSH"
+        # SSH verification has to be done here
+        known = ssh_knownhost_check(host, cert_p)
+        # appropriate warning already emitted (if any)
+        return known ? Consts.CERT_ACCEPT : Consts.CERT_REJECT
+    end
+    # unclear what other transports can occur here, but they may
+    # not default to secure, so we should reject anything unknown
+    return Consts.CERT_REJECT
+end
+
+# SSH known host checking
+struct CertHostKey
+    parent :: Cint
+    mask   :: Cint
+    md5    :: NTuple{16,UInt8}
+    sha1   :: NTuple{20,UInt8}
+    sha256 :: NTuple{32,UInt8}
+end
+
+struct KnownHost
+    magic :: Cuint
+    node  :: Ptr{Cvoid}
+    name  :: Ptr{Cchar}
+    key   :: Ptr{Cchar}
+    type  :: Cint
+end
+
+function ssh_knownhost_check(host::AbstractString, key_hashes_p::Ptr{Cvoid})
+    key_hashes = unsafe_load(convert(Ptr{CertHostKey}, key_hashes_p))
+    if key_hashes.mask & (Consts.CERT_SSH_SHA1 | Consts.CERT_SSH_SHA256) == 0
+        verify_host_error("SSH host verification: no secure certificate hash available for `$host`, cannot verify server identity.")
+        return false
+    end
+    session = @ccall libssh2_session_init_ex(
+        C_NULL :: Ptr{Cvoid},
+        C_NULL :: Ptr{Cvoid},
+        C_NULL :: Ptr{Cvoid},
+        C_NULL :: Ptr{Cvoid},
+    ) :: Ptr{Cvoid}
+    files = [joinpath(homedir(), ".ssh", "known_hosts")]
+    for file in files
```

Comment:
It looks like you're using the loop as goto alternative. I'd probably just use a goto as I think it's cleaner

Reply:
Oh I missed this comment: https://github.com/JuliaLang/julia/pull/38580#issuecomment-734052375. Carry on